fix(ivy): injecting incorrect provider when re-providing injectable with useClass

crisbeto opened the pull request:
If an injectable has a `useClass`, Ivy injects the token in `useClass`, rather than the original injectable, if the injectable is re-provided under a different token. The correct behavior is that it should inject the re-provided token, no matter whether it has `useClass`.

Fixes #34110.

kara commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -496,7 +497,8 @@ export function providerToFactory(
       } else {
-        return injectableDefOrInjectorDefFactory(classRef);
+        const classFactory = injectableDefOrInjectorDefFactory(classRef);
+        factory = () => classFactory(classRef);
```
This solution works well if the `classRef` doesn't use DI, but if `FooService` (token) and `BarService` (useClass in injectable def) have different DI signatures, we'll end up using `BarService`'s DI setup when instantiating `FooService`. So that case would be broken. Can we add a test for it?

In terms of a fix, it seems like the root problem is that we're going through the injectable def by default for anything that's explicitly provided. It seems like provider factories should not be doing this. Is there a reason we can't remove the `injectableDefOrInjectorDefFactory` call and replace it with `getFactoryDef` + the `getUndecoratedInjectableFactory` logic?

kara reviewed (changes requested):
Thanks for picking this up! Some Qs below

crisbeto committed:
fix(ivy): injecting incorrect provider when re-providing injectable with useClass

If an injectable has a `useClass`, Ivy injects the token in `useClass`, rather than the original injectable, if the injectable is re-provided under a different token. The correct behavior is that it should inject the re-provided token, no matter whether it has `useClass`.

Fixes #34110.

crisbeto commented:
Sorry for the delay @kara, the feedback has been addressed. Also you were right about the injected provider using the DI setup from the original class. I've implemented your comment on using the `getFactoryDef`, but I had to wrap the assertions in the tests with `if (ivyEnabled)`, because ViewEngine still has the behavior where it uses the wrong DI setup.

kara commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -1066,6 +1066,95 @@ describe('di', () => {
+      }
+    });
+  });
+
```
I think we're still missing a test for the explicit type provider case, e.g.

```
providers: [FooService]
```

kara commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -496,7 +497,8 @@ export function providerToFactory(
       } else {
-        return injectableDefOrInjectorDefFactory(classRef);
+        const classFactoryDef = getFactoryDef(classRef);
+        return classFactoryDef || injectableDefOrInjectorDefFactory(classRef);
```
I can't comment on the line, but for type providers, it seems like we would also want to update line 482. Otherwise, if you provide using `providers: [FooService]` pattern, the injectable def would still be used by default.

crisbeto committed:
fixup! fix(ivy): injecting incorrect provider when re-providing injectable with useClass

crisbeto commented:
The latest feedback has been addressed @kara.

kara reviewed (approved):
LGTM

kara commented:
[presubmit](http://test/OCL:297229463:BASE:297229506:1582676298616:fa96c17c)

kara commented:
merge-assistance: my approval should be sufficient since there are only changes in core files
